docs/design/2018-08-29-new-planner.md

```diff
@@ -0,0 +1,422 @@
+		}
+
+		eraseCur := findMoreEquiv(curExpr, g)
+		if eraseCur {
+			curGroup.erase(cur)
+		} else {
+			g.explored = g.explored && curExpr.explored
+		}
+	}
+}
+```
+
+The child of `GroupExpr` is `Group`. There are many candicate child expressions
+for a group expression. All the possible expressions have to be enumerated to
+check whether a group expression matches a transformation rule and apply the
+rule on that expression once matched.
+
+To conveniently enumerate all the equivalent expressions for a group expression,
+the `ExprIter` is introduced:
+
+```go
+// ExprIter enumerates all the equivalent expressions in the group according to
+// the expression pattern.
+type ExprIter struct {
+	// The group and ordinal field solely identify a group expression.
+	group   *Group
+	ordinal int
+
+	// operand is the node of the pattern tree. The operand type of the group
+	// expression must be matched with it, otherwise the group expression is
+	// ignored during the iteration.
+	operand int
+
+	// children is used to iterate the child expressions.
+	children *ExprIter
+}
+```
+
+To simplify and reduce the work of enumeration, we only consider the pattern
+that matches a rule. The pattern is a tree-like data structure which identifies
+the operator type for every node in the expression tree:
+
+```go
+type Pattern struct {
+	operand  int
+	children []*Pattern
+}
+```
+
+The `ExprIter` should be created with a `GroupExpr` and the `Pattern` of a rule:
+
+```go
+func NewExprIter(g *GroupExpr, p *Pattern) *ExprIter
+```
+
+With the help of the above components, the pseudo code of the function
+`findMoreEquiv` is:
+
+```go
+// Find and apply the matched transformation rules.
+func findMoreEquiv(cur *GroupExpr, curGroup *Group) (eraseCur bool) {
+	for _, rule := range GetTransformationRules(cur.exprNode) {
+		// Create a binding of the current group expression and the pattern of
+		// the transformation rule to enumerate all the possible expressions.
+		exprIter := NewExprIter(cur, rule.getPattern())
+		for exprIter.Next() {
+			if !rule.match(exprIter) {
+				continue
+			}
+
+			newExpr, erase, err := rule.onTransform(exprIter)
+			eraseCur = eraseCur || erase
+			if !curGroup.insert(newExpr) {
+				continue
+			}
+
+			// If the new group expression is successfully inserted into the
+			// current group, we mark the group expression and the group as
+			// unexplored to enable the exploration on the new group expression
+			// and all the antecedent groups.
+			newExpr.explored = false
+			curGroup.explored = false
+		}
+	}
+	return eraseCur
+}
+```
+
+According to the former discussion, the interface of the transformation rule can
+be defined as:
+
+```go
+type transformation interface {
+	getpattern() *Pattern
+	match(expr *ExprIter) (matched bool, err error)
+	onTransform(old *ExprIter) (new *GroupExpr, eraseOld bool, err error)
+}
+```
+
+At the very beginning, there is only one group expression in a `Group`, after
```

Comment:
At the very beginning, there is only one group expression in a **`Group`. After**